### ISSUE TITLE
wasm: musttail call must precede a ret with an optional bitcast

### ISSUE THREAD

**Issue by FrankReh** [tinygo-org/tinygo#1790]
```
% tinygo build -o try.wasm -target wasm
musttail call must precede a ret with an optional bitcast
  musttail call fastcc void @runtime.runqueuePushBack(%"internal/task.Task"* nonnull %parentHandle.reload1), !dbg !1172
error: GC pass caused a verification failure
```
Version reports
```
tinygo version 0.17.0 darwin/amd64 (using go version go1.16.3 and LLVM version 11.1.0)
```
Problem does not manifest when -opt=1 is used. Problem also reproduced from the latest dev branch.

Ran into this when building for wasm where a js.FuncOf was being called. Reduced the problem to find it is some combination of calls from the syscall/js package which I reduced further to a stand alone main.go file below. But the problem goes away if the sync.Lock call is removed so that sync package I felt I had to keep in the reduction. Many attempts at reducing the main.go further also made the problem go away so I could not reduce it down to just one line that makes a difference. Sorry there are so many functions in this reproduction.

Removing the sync.Lock makes the problem go away. Also just removing the one call to panic makes the problem go away. The great job LLVM does at removing dead code though makes it hard to guess why a line does or does not affect tail call logic. Even just removing the *int field from the struct effects the outcome, even though the code doesn't use the field - guess it is the size of the struct.

```
// +build js,wasm

package main

import "sync"

const (
	T0 int = iota
	T1
	T2
)

var (
	mu sync.Mutex
	v1 = make1(1) // change 1 to 0, get another kind of crash
	v2 = v1.Method1()
)

func main() {
	_ = LockAndReturn()
}
func LockAndReturn() error {
	mu.Lock()
	mu.Unlock()
	return nil
}

type S1 struct {
	f1 int
	f2 *int
}

func make1(b byte) S1 {
	return S1{f1: (int(b))}
}

func (v S1) Method1() S1 {
	if t := v.Type(); !isObject(t) {
		panic(&ValueError{})
	}
	return makeValue(valueGet(0, ""))
}

func makeValue(r int) S1 {
	return S1{}
}

func valueGet(v int, p string) int

type ValueError struct {
	Method string
}

func (e *ValueError) Error() string {
	return "something" + e.Method
}

func isObject(t int) bool {
	return t == T1 || t == T2
}

func (v S1) Type() int {
	if v.f1 == 0 {
		return T0
	}
	return T2
}
```

**Comment by FrankReh** [comment]
This is the IR. I believe it comes from the coroutine lowering. WebAssembly can't do tail calls yet so it's odd a musttail is even making it into the generated code. It seems related to the returnTo coroutine intrinsics setup in coroutines.go. In the middle of this IR, is a branch and a get.throw.i3 and then a second label. Here we have a tail call to runtime.nilPanic and later a musttail call to runtime.runqueuePushBack. The coroutine lowering step seems to be adding musttail simply because it knows it is the last call of the function, before the 'ret' return instruction. 

For better or worse, I found I can use the use.SetTailCall(false) on all call sites and have the tail and musttail removed and the code seems to function correctly.

```

"(*sync.Mutex).Unlock.exit":                      ; preds = %store.next.i2, %if.then.i
  %parentHandle.reload.addr = getelementptr inbounds %main.LockAndReturn.Frame, %main.LockAndReturn.Frame* %FramePtr, i32 0, i32 2, !dbg !1199
  %6 = bitcast i8** %parentHandle.reload.addr to %"internal/task.Task"**, !dbg !1199
  %parentHandle.reload1 = load %"internal/task.Task"*, %"internal/task.Task"** %6, align 4, !dbg !1199
  call void @llvm.dbg.value(metadata %"internal/task.Task"* undef, metadata !1239, metadata !DIExpression()), !dbg !1243
  call void @llvm.dbg.value(metadata i8* undef, metadata !1242, metadata !DIExpression()), !dbg !1243
  call void @llvm.dbg.value(metadata %"internal/task.Task"* undef, metadata !1239, metadata !DIExpression()), !dbg !1245
  %7 = icmp eq %"internal/task.Task"* %parentHandle.reload1, null, !dbg !1246
  br i1 %7, label %gep.throw.i3, label %"(*internal/task.Task).returnTo.exit", !dbg !1246

gep.throw.i3:                                     ; preds = %"(*sync.Mutex).Unlock.exit"
  tail call fastcc void @runtime.nilPanic(), !dbg !1246
  unreachable, !dbg !1246

"(*internal/task.Task).returnTo.exit":            ; preds = %"(*sync.Mutex).Unlock.exit"
  %task.state.parent.reload.addr = getelementptr inbounds %main.LockAndReturn.Frame, %main.LockAndReturn.Frame* %FramePtr, i32 0, i32 3, !dbg !1199
  %8 = bitcast i8** %task.state.parent.reload.addr to i32*, !dbg !1199
  %task.state.parent.reload2 = load i32, i32* %8, align 4, !dbg !1199
  %9 = getelementptr inbounds %"internal/task.Task", %"internal/task.Task"* %parentHandle.reload1, i32 0, i32 3, i32 0, !dbg !1247
  call void @llvm.dbg.value(metadata i8* undef, metadata !1242, metadata !DIExpression()), !dbg !1248
  %10 = bitcast i8** %9 to i32*, !dbg !1248
  store i32 %task.state.parent.reload2, i32* %10, align 4, !dbg !1248
  call void @llvm.dbg.value(metadata %"internal/task.Task"* undef, metadata !1239, metadata !DIExpression()), !dbg !1249
  call void @llvm.dbg.value(metadata %"internal/task.Task"* undef, metadata !1191, metadata !DIExpression()), !dbg !1250
  musttail call fastcc void @runtime.runqueuePushBack(%"internal/task.Task"* nonnull %parentHandle.reload1), !dbg !1252
  ret void, !dbg !1199
}
```

Any ideas what coroutine options I could play with or learn more about to explain this behavior? I still don't know if the musttail call is really violating an LLVM rule, it seems LLVM put it there.

**Comment by FrankReh** [comment]
Oh I see the problem. There is a pass that adds a store instruction between the musttail and the ret. I had been printing the IR out too early. With hindsight, this seems like a silly error on my part - not outputting the IR right before the error was reported. I had assumed the mistake was having a musttail in the first place. Now it is likely the problem is with the store being inserted there - perhaps the step that added the store could have put it before the last tail call or the pass could have removed the musttail marker itself. Here is the offending IR and now it is clear why there is a complaint.

```
"(*internal/task.Task).returnTo.exit":            ; preds = %"(*sync.Mutex).Unlock.exit"
  %task.state.parent.reload.addr = getelementptr inbounds %main.LockAndReturn.Frame, %main.LockAndReturn.Frame* %FramePtr, i32 0, i32 3, !dbg !1171
  %16 = bitcast i8** %task.state.parent.reload.addr to i32*, !dbg !1171
  %task.state.parent.reload2 = load i32, i32* %16, align 4, !dbg !1171
  %17 = getelementptr inbounds %"internal/task.Task", %"internal/task.Task"* %parentHandle.reload1, i32 0, i32 3, i32 0, !dbg !1217
  call void @llvm.dbg.value(metadata i8* undef, metadata !1212, metadata !DIExpression()), !dbg !1218
  %18 = bitcast i8** %17 to i32*, !dbg !1218
  store i32 %task.state.parent.reload2, i32* %18, align 4, !dbg !1218
  call void @llvm.dbg.value(metadata %"internal/task.Task"* undef, metadata !1209, metadata !DIExpression()), !dbg !1219
  call void @llvm.dbg.value(metadata %"internal/task.Task"* undef, metadata !1163, metadata !DIExpression()), !dbg !1220
  musttail call fastcc void @runtime.runqueuePushBack(%"internal/task.Task"* nonnull %parentHandle.reload1), !dbg !1222
  store %runtime.stackChainObject* %0, %runtime.stackChainObject** @runtime.stackChainStart, align 4, !dbg !1171
  ret void, !dbg !1171
}
```

**Comment by FrankReh** [comment]
```
		// Make sure this stack object is popped from the linked list of stack
		// objects at return.
		for _, ret := range returns {
			builder.SetInsertPointBefore(ret)
			builder.CreateStore(parent, stackChainStart)
		}
```

**Comment by FrankReh** [comment]
The musttail marker can be turned into a tail marker to solve this. Or the insertion point could be put before the call that has the musttail on it. Or if it is known that this is only used for targets that don't support tail calls any, all tail and musttail markers can be removed. I've tried all three approaches and they all solve the problem from my perspective.

**Comment by fgsch** [comment]
Also running into this with the following code (edited to show the problem):

```
package main

import (
	"time"
	"fmt"
)

func main() {
	c := make(chan string, 1)
	for i := 0; i < cap(c); i++ {
		go func(i int) {
			time.Sleep(time.Millisecond)
			fmt.Printf("goroutine %d\n", i)
		}(i)
	}
}
```

It builds fine with the leaking gc.

**Comment by fgsch** [comment]
Moved from #1979 :

```
package main

import "sync"

var once sync.Once

func bug() *int {
	once.Do(func() {})
	i := 0
	return &i
}

func main() {
	bug()
}
```

**Comment by aykevl** [comment]
Here is a fix: https://github.com/tinygo-org/tinygo/pull/2022

**Comment by deadprogram** [comment]
Released with 0.20.0 so now closing. Thank you!
